internal/observability/call.go

```diff
@@ -150,47 +175,91 @@ func (c call) endStats(elapsed time.Duration, err error, isApplicationError bool
 
 	isStatus := yarpcerrors.IsStatus(err)
 
-	// If the error is not a yarpcerrors.Status, we cannot determine a code or
-	// fault. Assume that these application errors are the caller's fault and emit
-	// a generic error tag.
-	if isApplicationError && !isStatus {
+	if isStatus {
+		status := yarpcerrors.FromError(err)
+		errCode := status.Code()
+		c.endStatsFromFault(elapsed, errCode, errCode.String())
+
+		return
+	}
+
+	if isApplicationError {
+		if applicationFailure != nil &&
+			applicationFailure.Code != nil {
+			errorName := applicationFailure.Name
+
+			if errorName == "" {
+				errorName = applicationFailure.Code.String()
+			}
+
+			c.endStatsFromFault(elapsed, *applicationFailure.Code, errorName)
+
+			return
+		}
+
+		// It is an application error but not a Status and no YARPC Code is found.
+		// Assume it's a caller's fault and emit generic error data.
 		c.edge.callerErrLatencies.Observe(elapsed)
-		if counter, err := c.edge.callerFailures.Get(_error, "application_error"); err == nil {
+
+		// The metrics system translates an empty string into "default".
+		// Better to instead specify "unknown-error", since the tag will always be there anyway.
+		appErrorNameTag := _unknownError
```

Comment:
This could be confused with [CodeUnknown](https://sourcegraph.com/github.com/yarpc/yarpc-go/-/blob/yarpcerrors/codes.go#L145), where the string there is "unknown".

Maybe we can use something like "__not_set__"?